Progress towards possibility of variable-sized segments

jaschiu opened the pull request:
Idea is that to support variable-sized segments, one big problem is to analyze the map for the space available in front of the player. This PR will try to solve it with a pure function that will calculate possible dimensions for potential segments based on given data. This means, the JavaScript will also need to be modified to provide more data. I have only wrote detailed documentation of this function as of now, more commits of actual code will come soon.

Commit:
[DOCUMENTATION]: Function that calculates possible segment dimensions

Commit:
[WIP]: Half of base code

pcrain commented:
Looks good so far! 😃 Will look over again and hopefully merge once you've finished the implementation.

Commit:
[WIP]: Finished possible_dimensions framework

[TODO]: Testing and debugging

Commit:
[MERGE]

Commit:
[WIP]: One test case passed
[UPDATE]: Changed directory structure to be more organized
[TODO]: More test cases and debugging

Commit:
Merge branch 'main' of https://github.com/cs240-illinois/sp22_cs240_infinite-maze-shared

Commit:
Merge branch 'main' of https://github.com/cs240-illinois/sp22_cs240_infinite-maze-shared

wadefagen commented on .gitignore:
```diff
@@ -1,4 +1,163 @@
-*/__pycache__/*
-*.pytest_cache/*
+<<<<<<< HEAD
```
This has git merge markers in the file itself -- please fix these

wadefagen reviewed (changes requested):
Noticed some merge markers in a quick glance at this code

Commit:
[WIP]: Lots of debugging, will add more test cases

[TODO]: Integration with middleware

Commit:
[FIX]: removed merge artifact inside .gitignore

Commit:
[UPDATE]: possible_dimensions should work now

[TODO]: Middleware integration

Commit:
[WIP]: More efficient code

[TODO]: possible_dimensions

Commit:
[UPDATE]: Relative coordinate function wrapper finished

[TODO]: Middleware integration

jasyip commented:
Main function is finished. I can try to analyze and modify the middleware/frontend a bit to send more data and modify which MGs are chosen, but may take a few days without help from other people on this issue.

Commit:
[UPDATE]: Fixed glitch that bypassed min_possible_len at
entrance_direction boundary

Commit:
[WIP]: A bit of middleware modification

jasyip commented:
Modifications needed for middleware:

- requests to middleware's `/generateSegment` should include the direction the player intended to move in.
- the data structure storing the global segment stores each individual coordinate as keys. I would have to convert it to a collection of square boundaries, wasting computational resources, so why not store the square boundaries *as* the keys?
- MGs will have to incorporate more data, such as the offsets of all of their exits on each side, hopefully it's not too much of a drastic change
- may have to make the random MG chooser more complex?

Commit:
[UPDATE]: Changed possible_dimensions return value

wadefagen commented:
> Edit: I also fixed my .gitignore having a merge artificat but I don't know how to get the Changes Requested notification to go away

On the reviewer area, you should be able to press the "refresh" button next to the name of the person who requested the change:
![image](https://user-images.githubusercontent.com/7505610/166159089-587bdc62-a740-44b4-a7de-38322e03e5d1.png)

jasyip commented:
> > Edit: I also fixed my .gitignore having a merge artificat but I don't know how to get the Changes Requested notification to go away
> 
> On the reviewer area, you should be able to press the "refresh" button next to the name of the person who requested the change: ![image](https://user-images.githubusercontent.com/7505610/166159089-587bdc62-a740-44b4-a7de-38322e03e5d1.png)

Done.

I'm going to try to spend some time figuring out how to modify some stuff in the middleware right now.

Commit:
Merge branch 'main' of https://github.com/cs240-illinois/sp22_cs240_infinite-maze-shared into dev

Commit:
Merge branch 'main' of https://github.com/cs240-illinois/sp22_cs240_infinite-maze-shared into dev

Commit:
[WIP]:
- Tackled how it would call the function
- need to change a lot of the middleware

jasyip commented:
I changed `app.py` `generateSegment` to make use of the function, but a lot of the middleware needs to have a different structure. I may not be able to finish this by today, but other people can help out.
